src/kOS/Suffixed/ResourceTransferValue.cs

```diff
@@ -0,0 +1,387 @@
+
+            var transferGoal = CalculateTransferGoal(toParts);
+
+            double pulledAmount = PullResources(fromParts, transferGoal);
+
+            PutResources(toParts, pulledAmount);
+
+            transferedAmount += pulledAmount;
+
+            if (Status == TransferManager.TransferStatus.Transferring)
+            {
+                StatusMessage = string.Format("Transfered: {0}", transferedAmount);
+            }
+            SafeHouse.Logger.Log("TRANSFER WORK: Have now transfered: " + transferedAmount);
+        }
+
+        /// <summary>
+        /// Transfers resources into the specified parts 
+        /// </summary>
+        /// <param name="parts">All of the recipient parts</param>
+        /// <param name="pulledAmount">the aggregate amount we are seeking to add to the parts</param>
+        private void PutResources(IList<global::Part> parts, double pulledAmount)
+        {
+            var retries = 0;
+            var evenShare = pulledAmount/parts.Count;
+
+            var remaining = pulledAmount;
+            while (remaining > 0.0001)
+            {
+                if (retries > 10)
+                {
+                    MarkFailed("Error in putting resource with " + remaining + " remaining.");
+                    break;
+                }
+                foreach (var part in parts)
+                {
+                    var resource = part.Resources.Get(resourceInfo.id);
+                    if (resource == null) continue;
+
+                    var transferAmount = Math.Min(remaining, evenShare);
+
+                    SafeHouse.Logger.Log(string.Format("TRANSFER WORK: {0} PUT AMOUNT: {1}", part.flightID, transferAmount));
+
+                    remaining += part.TransferResource(resource.info.id, transferAmount);
+
+                    SafeHouse.Logger.Log(string.Format("TRANSFER WORK: {0} PUT REMAINING: {1}", part.flightID, remaining));
+                }
+                retries++;
+            }
+        }
+
+        /// <summary>
+        /// Requests the resource from all parts in the collection
+        /// </summary>
+        /// <param name="parts">All of the donor parts</param>
+        /// <param name="transferGoal">the aggregate amount we are seeking to remove from parts</param>
+        /// <returns>the amount we were successful at pulling</returns>
+        private double PullResources(IList<global::Part> parts, double transferGoal)
+        {
+            double toReturn = 0.0;
+            var availableResources = CalculateAvailableResource(parts);
+            foreach (var part in parts)
+            {
+                var resource = part.Resources.Get(resourceInfo.id);
+                if (resource == null) continue;
+
+                var thisPartsPercentage = resource.amount/availableResources;
+
+                // Throttle the transfer
+                var thisPartsShare = transferGoal*thisPartsPercentage;
+                var thisPartsRate = resource.maxAmount*RESOURCE_SHARE_PER_UPDATE;
```

Comment:
Yeah, i couldnt find the actual rate so i took a guess and it seemed ok. we can of course always change it.